password-cli/app/Main.hs

```diff
@@ -0,0 +1,177 @@
+{-# LANGUAGE DuplicateRecordFields #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RecordWildCards #-}
+
+module Main (main) where
+
+import Control.Monad (unless, void)
+import qualified Data.Password.Argon2 as Argon2
+import Data.Password.Bcrypt (PasswordCheck (..))
+import qualified Data.Password.Bcrypt as Bcrypt
+import qualified Data.Password.PBKDF2 as PBKDF2
+import qualified Data.Password.Scrypt as Scrypt
+import Data.Password.Types
+import Data.Text (Text)
+import qualified Data.Text.IO as T
+import Options.Applicative
+import System.Exit (exitFailure)
+import System.IO (IOMode(ReadMode), stdin, withFile)
+
+main :: IO ()
+main = execParser cliOpts >>= runCmd
+
+data Cmd
+  = HashCmd HashOpts
+  | CheckCmd CheckOpts
+  | HelpCmd (Maybe String)
+
+data HashOpts = HashOpts {
+    password :: Either FilePath Bool,
+    hashAlgorithm :: HashAlgorithm
+  }
+
+data CheckOpts = CheckOpts {
+    hash :: Either FilePath Text,
+    password :: Either FilePath Bool,
+    checkAlgorithm :: CheckAlgorithm
+  }
+
+data HashAlgorithm
+  = PBKDF2HashAlgo PBKDF2.PBKDF2Params
+  | BcryptHashAlgo Int
+  | ScryptHashAlgo Scrypt.ScryptParams
+  | Argon2HashAlgo Argon2.Argon2Params
+
+data CheckAlgorithm
+  = PBKDF2CheckAlgo
+  | BcryptCheckAlgo
+  | ScryptCheckAlgo
+  | Argon2CheckAlgo
+
+cliOpts :: ParserInfo Cmd
+cliOpts = info commandsParser (fullDesc <> header "password CLI usage -")
+  where
+    commandsParser :: Parser Cmd
+    commandsParser =
+      subparser
+        ( command
+            "hash"
+            (info (HashCmd <$> hashOptsParser) (progDesc "hash password"))
+            <> command
+              "check"
+              (info (CheckCmd <$> checkOptsParser) (progDesc "check hashed password"))
+            <> command "help" (info (HelpCmd <$> optional (argument str (metavar "COMMAND")) <**> helper) (progDesc "Show command help"))
+        )
+        <**> helper
+    hashOptsParser :: Parser HashOpts
+    hashOptsParser =
+      HashOpts
+        <$> (   (Left <$> option str (metavar "PASSWORD-FILE" <> long "password-file"))
```

Comment:
I prefer the parsing of options on their own over the `Either l r` parsing way of doing it, because in the main program, you can then check if conflicting options have been set, abort and report that the options given are conflicting, instead of silently dropping one.

The other option is to explicitly state in the description of the overriding function that using it will override other options.

Seeing this package (and consequently this CLI) is all about security and validation. I'd like to keep it straight forward and just collect all options, and validate them before running to give good error messages to the user.

Reply:
Actually, it does not silently drop it:

```
exe hash argon2 --password-file /tmp/pw -q
Invalid option `-q'

Usage: password-cli COMMAND
```

Moreover it's properly documented:

```
$ exe --help hash
Usage: password-cli hash [--password-file PASSWORD-FILE | (-q|--quiet)] COMMAND
```

Reply:
But what if the `--password-file` was the mistake, and now the user suddenly hears `-q` is not a valid option. That's not a good user experience. It'd be nicer to hear "Can't use `--password-file` and `-q` at the same time".

Though in this case, `-q` should probably just be a global flag that prevents any logging/reporting output in general, and accept the option even if the command is not going to log/report anything.